core/lib/server/pub.ts

```diff
@@ -589,3 +600,574 @@ export const deletePub = async (pubId: PubsId) =>
+	fieldSlugs?: string[];
+} & GetManyParams;
+
+type PubIdOrPubTypeIdOrStageIdOrCommunityId =
+	| {
+			pubId: PubsId;
+			pubTypeId?: never;
+			stageId?: never;
+			communityId?: never;
+	  }
+	| {
+			pubId?: never;
+			pubTypeId: PubTypesId;
+			stageId?: never;
+			communityId?: never;
+	  }
+	| {
+			pubId?: never;
+			pubTypeId?: never;
+			stageId: StagesId;
+			communityId?: never;
+	  }
+	| {
+			pubId?: never;
+			pubTypeId?: never;
+			stageId?: never;
+			communityId: CommunitiesId;
+	  };
+
+/**
+ * Only add the `children` if the `withChildren` option has not been set to `false
+ */
+type MaybePubChildren<Options extends GetPubsWithRelatedValuesAndChildrenOptions> =
+	Options["withChildren"] extends false
+		? { children?: undefined }
+		: { children: ProcessedPub<Options>[] };
+
+/**
+ * Only add the `stage` if the `withStage` option has not been set to `false
+ */
+type MaybePubStage<Options extends GetPubsWithRelatedValuesAndChildrenOptions> =
+	Options["withStage"] extends true ? { stage: Stages } : { stage?: undefined };
+
+/**
+ * Only add the `pubType` if the `withPubType` option has not been set to `false
+ */
+type MaybePubPubType<Options extends GetPubsWithRelatedValuesAndChildrenOptions> =
+	Options["withPubType"] extends true ? { pubType: PubTypes } : { pubType?: undefined };
+
+type MaybeWithRelatedPub<Options extends GetPubsWithRelatedValuesAndChildrenOptions> =
+	Options["withRelatedPubs"] extends false
+		? { relatedPub?: undefined }
+		: { relatedPub: ProcessedPub<Options>[] };
+
+/**
+ * Those options of `GetPubsWithRelatedValuesAndChildrenOptions` that affect the output of `ProcessedPub`
+ *
+ * This way it's more easy to specify what kind of `ProcessedPub` we want as e.g. the input type of a function
+ *
+ **/
+type MaybeOptions = Pick<
+	GetPubsWithRelatedValuesAndChildrenOptions,
+	"withChildren" | "withRelatedPubs" | "withPubType" | "withStage"
+>;
+
+export type ProcessedPub<Options extends MaybeOptions = {}> = {
+	id: PubsId;
+	stageId: StagesId | null;
+	communityId: CommunitiesId;
+	pubTypeId: PubTypesId;
+	parentId: PubsId | null;
+	/**
+	 * An array of values for the pub.
+	 */
+	values: ({
+		id: PubValuesId;
+		fieldId: PubFieldsId;
+		value: unknown;
+		relatedPubId: PubsId | null;
+		createdAt: Date;
+		updatedAt: Date;
+		/**
+		 * Information about the field that the value belongs to.
+		 */
+		field: {
+			schemaName: CoreSchemaType;
+			slug: string;
+		};
+	} & MaybeWithRelatedPub<Options>)[];
+	createdAt: Date;
+	/**
+	 * The `updatedAt` of the latest value, or of the pub if the pub itself has a higher `updatedAt` or if there are no values
+	 *
+	 * We do this because the Pub itself is rarely if ever changed over time.
+	 * TODO: Possibly add the `updatedAt` of `PubsInStages` here as well?
+	 * At time of writing (2024/11/04) I don't think that table has an `updatedAt`.
+	 */
+	updatedAt: Date;
+} & MaybePubChildren<Options> &
+	MaybePubStage<Options> &
+	MaybePubPubType<Options>;
```

Comment:
maybe a bit ugly, but this is the only way i found to do roughly this pseudo type code

```ts
if( options.withPubType === true ) {
	return { pubType: PubType }
} else {
 	return { pubType?: never }
}
```

This is bc it's not possible to do "conditional optionality" in typescript, ie you cant say "only make this type optional for a certain condition". You need to do some messed up stuff for that.

Reply:
I think optional `never` is preferable over just leaving it out, because it allows you to do 

```ts
if(pub.pubType){
}
```
rather than having to do

```ts
if('pubType' in pub && pub.pubType){
}
```